dataset/src/main/scala/frameless/functions/Udf.scala

```diff
@@ -0,0 +1,80 @@
+package frameless.functions
+
+import frameless.{TypedColumn, TypedEncoder}
+import org.apache.spark.sql.Column
+import org.apache.spark.sql.functions.{udf => sudf}
+
+import scala.reflect.runtime.universe._
+
+trait Udf {
+  def udf[T, A: TypeTag, R: TypeTag](
+     f: A => R)
+  (implicit
+     aenc: TypedEncoder[A],
+     reng: TypedEncoder[R]):
+  (TypedColumn[T, A] => TypedColumn[T, R]) =
+    (u: TypedColumn[T, A]) => {
+      val uf = sudf(f)
+      val x: Column = uf(u.untyped)
+      new TypedColumn[T,R](x)
+    }
+
+  def udf[T, A1: TypeTag, A2: TypeTag, R: TypeTag](
+     f: (A1,A2) => R)
+  (implicit
+     a1enc: TypedEncoder[A1],
+     a2enc: TypedEncoder[A2],
+     reng: TypedEncoder[R]):
+  ( (TypedColumn[T, A1],TypedColumn[T, A2]) => TypedColumn[T, R]) =
+    (u1: TypedColumn[T, A1], u2: TypedColumn[T, A2]) => {
+      val uf = sudf(f)
+      val x: Column = uf(u1.untyped, u2.untyped)
+      new TypedColumn[T,R](x)
+    }
+
+  def udf[T, A1: TypeTag, A2: TypeTag, A3: TypeTag, R: TypeTag](
+  f: (A1,A2,A3) => R)
+  (implicit
+   a1enc: TypedEncoder[A1],
+   a2enc: TypedEncoder[A2],
+   a3enc: TypedEncoder[A3],
+   reng: TypedEncoder[R]):
+  ( (TypedColumn[T, A1],TypedColumn[T, A2],TypedColumn[T, A3]) => TypedColumn[T, R]) =
+    (u1: TypedColumn[T, A1], u2: TypedColumn[T, A2], u3: TypedColumn[T, A3]) => {
+      val uf = sudf(f)
+      val x: Column = uf(u1.untyped, u2.untyped, u3.untyped)
+      new TypedColumn[T,R](x)
+    }
+
+  def udf[T, A1: TypeTag, A2: TypeTag, A3: TypeTag, A4: TypeTag, R: TypeTag](
+   f: (A1,A2,A3,A4) => R)
+   (implicit
+    a1enc: TypedEncoder[A1],
+    a2enc: TypedEncoder[A2],
+    a3enc: TypedEncoder[A3],
+    a4enc: TypedEncoder[A4],
+    reng: TypedEncoder[R]):
+  ( (TypedColumn[T, A1],TypedColumn[T, A2],TypedColumn[T, A3],TypedColumn[T, A4]) => TypedColumn[T, R]) =
+    (u1: TypedColumn[T, A1], u2: TypedColumn[T, A2], u3: TypedColumn[T, A3], u4: TypedColumn[T, A4]) => {
+      val uf = sudf(f)
+      val x: Column = uf(u1.untyped, u2.untyped, u3.untyped, u4.untyped)
+      new TypedColumn[T,R](x)
+    }
+
+  def udf[T, A1: TypeTag, A2: TypeTag, A3: TypeTag, A4: TypeTag, A5: TypeTag, R: TypeTag](
+    f: (A1,A2,A3,A4,A5) => R)
+    (implicit
+     a1enc: TypedEncoder[A1],
+     a2enc: TypedEncoder[A2],
+     a3enc: TypedEncoder[A3],
+     a4enc: TypedEncoder[A4],
+     a5enc: TypedEncoder[A5],
+     reng: TypedEncoder[R]):
+  ( (TypedColumn[T, A1],TypedColumn[T, A2],TypedColumn[T, A3],TypedColumn[T, A4],TypedColumn[T, A5]) => TypedColumn[T, R]) =
+    (u1: TypedColumn[T, A1], u2: TypedColumn[T, A2], u3: TypedColumn[T, A3], u4: TypedColumn[T, A4], u5: TypedColumn[T, A5]) => {
```

Comment:
AFAIK there is no need to specify types here, they would be inferred by the compiler because function signature already has types.